File: conformance/run-1/components/broadcast.py
"""Public Python broadcast composition; the host owns delivery and isolation."""

import asyncio

from jiggy.flow import ChannelReceiver, JsonValue, OperationError, RunContext, RunResult, handle


async def run(context: RunContext) -> RunResult:
    source = await context.channel(delivery="broadcast")
    early = await source.subscribe()
    await source.send.send("before-late")
    late = await source.subscribe()
    await source.send.send("after-late")
    await source.send.close()

    async def collect(receiver: ChannelReceiver) -> JsonValue:
        values: list[JsonValue] = []
        complete = True
        try:
            async for value in receiver:
                values.append(value)
        except OperationError as error:
            if error.code != "LAGGED":
                raise
            complete = False
        return {"start": receiver.start_sequence, "values": values, "complete": complete}

    first, second = await asyncio.gather(collect(early), collect(late), return_exceptions=True)
    if isinstance(first, BaseException):
        raise first
    if isinstance(second, BaseException):
        raise second
    return {"outcome": "done", "output": {"early": first, "late": second}}


handle(run)

File: conformance/run-0/components/broadcast.ts
import {
  handle,
  OperationError,
  type ChannelReceiver,
  type JsonValue,
} from '../../../packages/flow-sdk/src/index'

async function collect(receiver: ChannelReceiver) {
  const values: JsonValue[] = []
  let complete = true
  try {
    for await (const value of receiver) values.push(value)
  } catch (error) {
    if (!(error instanceof OperationError) || error.code !== 'LAGGED') throw error
    complete = false
  }
  return { start: receiver.startSequence, values, complete }
}

await handle(async (run) => {
  const source = await run.channel({ delivery: 'broadcast' })
  const early = await source.subscribe()
  await source.send.send('before-late')
  const late = await source.subscribe()
  await source.send.send('after-late')
  await source.send.close()
  const [first, second] = await Promise.all([collect(early), collect(late)])
  return { outcome: 'done', output: { early: first, late: second } }
})
